### PR TITLE
C#: SQL Injection improvements for SQLite.

### PR DESCRIPTION
**Pull request by michaelnebel**
The content on this PR is to address the issue: https://github.com/github/codeql-csharp-team/issues/172
Following the link to the original ticket it is worth noticing that
1. The project referenced on the ticket doesn't build, which could lead to the missing results.
2. The project referenced on the ticket is fork of another project which does build and
    - The claimed missing SQL injection is found by our query even though we don't directly support the SQLite framework.
    - The claimed missing Second Order SQL injection is in parts of the code that isn't included in the build. That being said, the Second Order SQL injection query will not find this.

In this PR we
1. Provide better support for the SQLite framework.
   - Explicit summaries for this Framework has been made, including Adapters (which was not supported before).
   - Stubs for the SQLite framework has been included.
   - Coding examples with weaknesses discovered by the SQL injection query are included.
2. FileStream and StreamReader summaries.
   - Flow summaries for FileStream and StreamReader has been included.
   - Example similar to the Second Order SQL injection has been provided and will no be caught as a first order SQL injection, if the filename is tainted.
4. FileStreams are now considered StoredFlowSources.
   - FileStreams are now considered stored flow sources.
   - Second Order SQL injection now catches an example similar to the one provided in the referenced ticket.

A change note will be added, when the exact scope of the PR is decided.

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll [hunk 1]
```diff
@@ -67,18 +68,46 @@ private class IDbCommandConstructionSinkModelCsv extends SinkModelCsv {
...
 }
 
-/** A construction of an `SqlDataAdapter` object. */
+/** Data flow for SqlCommand and friends. */
+private class SqlCommandSummaryModelCsv extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        // SqlCommand
+        "System.Data.SqlClient;SqlCommand;false;SqlCommand;(System.String);;Argument[0];Argument[Qualifier];taint;manual",
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll:86]: Why is the output for all constructors `Argument[Qualifier]` and not `ReturnValue`?

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll:86]: Constructors are speciel in the way that they are considered to be side effect based, ie. that they manipulate `this`. As far as I remember the reasoning behind this is related to inheritance or maybe I am mistaken?

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll:86]: For other constructors, like `Tuple<>`, we use `ReturnValue` (see https://github.com/github/codeql/blob/main/csharp/ql/lib/semmle/code/csharp/frameworks/System.qll#L1323).

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll:86]: Oh no - you are right. We need to fix all those manual written summaries - I don't think they work at all. Constructors are considered to be side effect based (ie. they manipulate an implicit this argument).
Fortunately, the model generator already handles constructors correctly.
Should I make a tech debt PR and start to fix all the hand written summaries (we (and by we  - I mean I) must have converted constructors incorrectly when removing the legacy data flow library))?

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll:86]: > Should I make a tech debt PR and start to fix all the hand written summaries

Yes, we should definitely fix those broken summaries. Thanks :-)

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Sql.qll:86]: As discussed offline, `Argument[Qualifier]` is arguably the right thing to write. Sorry about the confusion.

**Review by hvitved — CHANGES_REQUESTED** [review]
Unit tests are failing; otherwise LGTM.

**Comment by michaelnebel** [comment]
> Unit tests are failing; otherwise LGTM.

Yes, this is due to the rebase after the new stubs were generated. Will fix.
I will also add a release note and start DCA.

**Review by hvitved — DISMISSED** [review]

File: csharp/ql/src/change-notes/2022-08-10-sqlinjection-queries.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+category: minorAnalysis
+---
+* Added better support for the SQ Lite framework in the SQL injection query.
```
Comment (tamasvajk) [inline csharp/ql/src/change-notes/2022-08-10-sqlinjection-queries.md:4]: Nitpicking: There's an extra space in `SQ Lite`.

Comment (michaelnebel) [reply csharp/ql/src/change-notes/2022-08-10-sqlinjection-queries.md]: Thanks for catching it! Will update it!

**Comment by michaelnebel** [comment]
Lots of projects failed in DCA. Will re-execute DCA as there is an alerts diff on the projects that succeeded (which is not totally unexpected).

**Comment by michaelnebel** [comment]
Comments on the DCA execution
- There are no performance regressions.
- The failure of ASP.NET Core is most likely unrelated (it fails with out of memory - the same was the case for the nightly run)
- The new alerts are all for queries that use StoredFlowSource as a source of taint and the first step is opening a FileStream, which makes sense, since we are now considered a file a StoredFlowSource. Are these alerts something we want?

**Comment by hvitved** [comment]
> * The new alerts are all for queries that use StoredFlowSource as a source of taint and the first step is opening a FileStream, which makes sense, since we are now considered a file a StoredFlowSource. Are these alerts something we want?

AFAICT, there are no new results on WebGoat. Were we expecting that?

**Comment by michaelnebel** [comment]
> > * The new alerts are all for queries that use StoredFlowSource as a source of taint and the first step is opening a FileStream, which makes sense, since we are now considered a file a StoredFlowSource. Are these alerts something we want?
> 
> AFAICT, there are no new results on WebGoat. Were we expecting that?

No, we were not expecting more results.
(1) The first order SQL injection result was already found without any changes.
(2) The second SQL injection result will not be a found as the references code is not included in the compilation of the project.

**Comment by hvitved** [comment]
> the references code is not included in the compilation of the project

🤦

**Review by hvitved — APPROVED** [review]

**Pull request merged**